### PR TITLE
Ctx in parent callback

### LINKED ISSUE #210: Add contexts to callbacks in pipeline and bin
Pipeline's and Bin's callbacks should have contexts containing at least current playback state and their children (name, module, maybe options). Bin's callbacks should also have their pads data passed as elements do.

### PR DESCRIPTION
**Pull request by FelonEkonom**
fix #210 
TODO:
- [x] fix conflicts
- [x] update test bins and pipelines
- [x] refactor
- [x] spec generation
- [x] docs

### CODE REVIEW

**Comment by mat-hek** [comment]
please fix merge conficts

File: lib/membrane/core/callback_context.ex [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          require unquote(__MODULE__)
+          @behaviour unquote(__MODULE__)
+
+          @type t :: %__MODULE__{unquote_splicing(fields ++ default_fields)}
+
+          fields_names = unquote(Keyword.keys(fields))
+          default_fields_names = unquote(Keyword.keys(default_fields))
+
+          @enforce_keys Module.get_attribute(__MODULE__, :enforce_keys)
+                        ~> (&1 || fields_names)
```
Comment (bblaszkow06) [inline lib/membrane/core/callback_context.ex:41]: Module.get_attribute has 3rd argument `deafult` that you can use here

Comment (bblaszkow06) [inline lib/membrane/core/callback_context.ex:41]: `Bunch` is not "used" in this context, this works by accident

Comment (mat-hek) [reply lib/membrane/core/callback_context.ex:41]:
```
[
                      {:enforce_keys, [context: Membrane.Core.CallbackContext],
                       [
                         {:~>,
                          [
                            context: Membrane.Core.CallbackContext,
                            import: Bunch
                          ],
                          [
                            {:|>,
                             [
                               context: Membrane.Core.CallbackContext,
                               import: Kernel
                             ], [{...}, ...]},
                            {:++, [context: Membrane.Core.CallbackContext, ...],
                             [...]}
                          ]}
                       ]}
                    ]},
```

File: lib/membrane/bin.ex [hunk 2]
```diff
@@ -68,7 +52,7 @@ defmodule Membrane.Bin do
   `c:GenServer.init/1` callback.
   """
   @callback handle_init(options :: options_t) ::
-              {{:ok, [Membrane.Parent.Action.spec_action_t()]}, Bin.State.internal_state_t()}
+              {{:ok, [Membrane.Parent.Action.spec_action_t()]}, any}
```
[the remark this answers is not in the document]
Comment (mat-hek) [reply lib/membrane/bin.ex:55]: `Bin.State` is not public. I can do `state :: any` or declare `state_t` in `Membrane.Bin`

**Pull request merged**